File: Data Structures & Algorithms/permutations/submission-2.py
class Solution:
    def permute(self, nums: List[int],) -> List[List[int]]:
        res, sol = [], []
        used = [False] * len(nums)

        def backtrack(i):
            if (len(sol) == len(nums)):
                res.append(sol[:])
                return res
            
            for i in range(len(nums)):
                if not used[i]:
                    sol.append(nums[i])
                    used[i] = True
                    backtrack(i + 1)
                    sol.pop()
                    used[i] = False
        
        backtrack(0)
        return res
        
        

File: Data Structures & Algorithms/subsets-ii/submission-2.java
class Solution {
    private void backtrack(int[] nums, int i, List<Integer> subset, List<List<Integer>> res) {
        if (i == nums.length) {
            res.add(new ArrayList<>(subset));
            return;
        }

        subset.add(nums[i]);
        backtrack(nums, i + 1, subset, res);
        subset.remove(subset.size() - 1);

        while (i + 1 < nums.length && nums[i] == nums[i + 1]) {
            i++;
        }

        backtrack(nums, i + 1, subset, res);
    }

    public List<List<Integer>> subsetsWithDup(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> subset = new ArrayList<>();

        backtrack(nums, 0, subset, res);
        return res;
    }
}
